### PR TITLE
Send warning and do not return error on group update if cluster is a PR secondary and RPC client is nil

### PR DESCRIPTION
**Pull request by elliesterner**
### Description
If a DR secondary is promoted to become a DR Primary as well as a PR secondary, the promotion will fail and the cluster will never become available. This is because of an ongoing issue with duplicate aliases. If vault detects duplicate aliases, it will merge them, causing vault to update it's groups with the new entity members. However, the newly promoted DR primary/PR secondary fails to persist the group change to storage because it's rpc client has not been initialized yet.

To avoid this failure, we check if `UpsertGroupInTxn` returns an ErrReadOnly error, and instead log a warning message.

[Jira](https://hashicorp.atlassian.net/browse/VAULT-28191)

### TODO only if you're a HashiCorp employee
- [x] **Labels:** If this PR is the CE portion of an ENT change, and that ENT change is
  getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
  instead of the old style `backport/x.x.x` labels.
- [x] **Labels:** If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).
- [ ] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [ ] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [ ] **RFC:** If this change has an associated RFC, please link it in the description.
- [ ] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: vault/identity_store_util.go
Comment (biazmoreira) [inline vault/identity_store_util.go:182]: Maybe I missed this, but when will it be updated, then? Do we need to update it in storage at all?

Comment (elliesterner) [reply vault/identity_store_util.go:182]: No in this particular case (when we get the ErrReadOnly) we are skipping the step to persist it to storage.

Comment (biazmoreira) [reply vault/identity_store_util.go:182]: Sorry, I might be missing something, but how and when will all the nodes update memdb correctly? I remember reading something about this code being here to fix another issue.

Comment (elliesterner) [reply vault/identity_store_util.go:182]: Don't apologize :) there are so many moving pieces and I appreciate the questions, I may be missing something myself. The [UpsertGroupInTxn] also updates the group in memdb [here](https://github.com/hashicorp/vault/blob/main/vault/identity_store_util.go#L1835). So in the case when we see the `ErrReadOnly` we are not rolling back the entire transaction, so even if the storage update fails the memdb update will not.

Comment (biazmoreira) [reply vault/identity_store_util.go:182]: And when will the storage pick up the changes? I think what I mean is: when will storage and memdb synchronize? After the WALs get to the cluster replica/node replica? 

Are we aware/ok with the race conditions this might cause? For example, issuing an update/delete to the group right after the cluster has initialized and if the WALs aren't up to date?

**Review by biazmoreira — APPROVED** [review]

**Review by banks — APPROVED** [review]
The Ent part of this is passing CI now with the additional tests for this change.

I have some minor readability comments inline. All are optional so I want to unblock the PR but if you agree with the suggestions feel free to fix those up before merge!

File: vault/identity_store_util.go
Comment (elliesterner) [reply vault/identity_store_util.go:184]: I agree, I like the separate ifs.  But because we're not returning from the first if, won't we continue on to the second if as well and cause an error?

Comment (elliesterner) [reply vault/identity_store_util.go:184]: I could do something like: 
```

			if !errors.Is(err, logical.ErrReadOnly) {
				txn.Abort()
				return fmt.Errorf("failed to update group in memdb: %w", err)
			}

			if errors.Is(err, logical.ErrReadOnly) {
				// This is an imperfect solution to unblock customers who are running into
				// a readonly error during a DR failover (jira #28191). More specifically, there
				// are duplicate aliases in storage that are merged during loadEntities. Vault
				// attempts to remove these deleted entities from the group, but fails in the case
				// where the node is a PR secondary because the RPC client is not yet initialized
				// and the storage is read-only.
				i.logger.Warn("received a read only error while trying to upsert group to storage")
			}
			```

Comment (banks) [reply vault/identity_store_util.go:184]: > won't we continue on to the second if as well and cause an error?

I don't think so - because we set err to nil the next if will always be false and we'll continue exactly like there was no error at all.

Comment (elliesterner) [reply vault/identity_store_util.go:184]: ah yes! youre right, thank you for explaining! I like that too

**Pull request merged**
